File: ColorTempConverter.cpp
#include <iostream>
#include <cmath>
#include <string>

using namespace std;

void KelvinTempCalc();
int main() {
    cout << endl;
    cout << "Kelvin Color Temperature Calculator" << endl << endl;
    cout << "Enter the temperature in Kelvin, and it will output the RGB values. Type 'quit' at the end of the calculations to quit, or type anything else to calculate another value:" << endl << endl;
    std::string userInput;
    while (userInput != "quit") {
        KelvinTempCalc();
        cin >> userInput;
    }
    return 0;
}

void KelvinTempCalc() {
    int kelvinTemp;
    double red;
    double green;
    double blue;

    cout << "Enter Color Temperature in Kelvin: ";
    cin >> kelvinTemp;

    kelvinTemp = kelvinTemp /100;

    //Calculate Red
    if (kelvinTemp <= 66) {
      red = 255;
    }
    else {
      red = kelvinTemp - 60;
      red = 329.698727446 * pow(red, -0.1332047592);
      if (red < 0) {
        red = 0;
      }
      if (red > 255) {
        red = 255;
      }
    }

    //Calculate Green
    if (kelvinTemp <= 66) {
      green = kelvinTemp;
      green = 99.4708025861 * log(green) - 161.1195681661;
      if (green < 0) {
        green = 0;
      }
      if (green > 255) {
        green = 255;
      }
    }
    else {
      green = kelvinTemp - 60;
      green = 288.1221695283 * pow(green, -0.0755148492);
      if (green < 0) {
        green = 0;
      }
      if (green > 255) {
        green = 255;
      }
    }

    //Calculate blue
    if (kelvinTemp >= 66) {
      blue = 255;
    }
    else {
      if (kelvinTemp <= 19) {
        blue = 0;
      }
      else {
        blue = kelvinTemp - 10;
        blue = 138.5177312231 * log(blue) - 305.0447927307;
        if (blue < 0) {
          blue = 0;
        }
        if (blue > 255) {
          blue = 255;
        }
      }
    }

    cout << "RGB values are: " << red << "," << green << "," << blue << endl;
    red = red/255;
    green = green/255;
    blue = blue/255;

    cout << "In 0-1 range they are: " << red << "," << green << "," << blue << " ";
}

File: ColorTempConverter.py
import numpy as np

kelvinTemp = 1000

kelvinTemp = kelvinTemp / 100

def RestrictBounds(value):
    if value < 0:
        return 0
    if value > 255:
        return 255
    else:
        return value

def ConvertToDecimal(value):
    return value / 255

def CalculateRed(kelvinTemp):
    if kelvinTemp <= 66:
        return 255
    else:
        red = kelvinTemp - 60
        red = 329.698727446 * (red ** -0.1332047592)
        return RestrictBounds(red)

def RGBToHex(red, green, blue):
    rgb = (int(red), int(green), int(blue))
    return '%02x%02x%02x' % rgb

def CalculateGreen(kelvinTemp):
    if kelvinTemp <= 66:
        green = kelvinTemp
        green = 99.4708025861 * np.log(green) - 161.1195681661
        return RestrictBounds(green)
    else:
        green = kelvinTemp - 60
        green = 288.1221695283 * (green ** -0.0755148492)
        return RestrictBounds(green)

def CalculateBlue(kelvinTemp):
    if kelvinTemp >= 66:
        return 255
    else:
        if kelvinTemp <= 19:
            return 0
        else:
            blue = kelvinTemp - 10
            blue = 138.5177312231 * np.log(blue) - 305.0447927307
            return RestrictBounds(blue)

red = CalculateRed(kelvinTemp)
green = CalculateGreen(kelvinTemp)
blue = CalculateBlue(kelvinTemp)

red_decimal = ConvertToDecimal(red)
green_decimal = ConvertToDecimal(green)
blue_decimal = ConvertToDecimal(blue)

hexidecimal_value = RGBToHex(red, green, blue)

print("rgb (0-255) values are: ", red, ",", green, ",", blue)
print("rgb (0-1)   values are: ", red_decimal, ",", green, ",", blue)
print("Hexidecimal value is    ", hexidecimal_value)
